src/backend/distributed/operations/rebalancer_placement_isolation.c

```diff
@@ -0,0 +1,474 @@
+	int32 nodeGroupId;
+
+	/*
+	 * Whether given node is allowed to have any shards.
+	 *
+	 * This is not just WorkerNode->shouldHaveShards but also takes into account
+	 * whether the node is being drained.
+	 */
+	bool shouldHaveShards;
+
+	/*
+	 * Whether given node is allowed to separate any shard placement groups.
+	 */
+	bool allowedToSeparateAnyPlacementGroup;
+
+	/*
+	 * Shard placement group that is assigned to this node to be separated
+	 * from others in the cluster.
+	 *
+	 * NULL if no shard placement group is assigned yet.
+	 */
+	ShardPlacementGroup *assignedPlacementGroup;
+} NodeToPlacementGroupHashEntry;
+
+/*
+ * Routines to prepare a hash table where each entry is of type
+ * NodeToPlacementGroupHashEntry.
+ */
+static void NodeToPlacementGroupHashInit(HTAB *nodePlacementGroupHash,
+										 List *activeWorkerNodeList,
+										 List *rebalancePlacementList,
+										 WorkerNode *drainWorkerNode);
+static void NodeToPlacementGroupHashAssignNodes(HTAB *nodePlacementGroupHash,
+												List *activeWorkerNodeList,
+												List *rebalancePlacementList,
+												FmgrInfo *shardAllowedOnNodeUDF);
+static bool NodeToPlacementGroupHashAssignNode(HTAB *nodePlacementGroupHash,
+											   int32 nodeGroupId,
+											   ShardPlacement *shardPlacement,
+											   FmgrInfo *shardAllowedOnNodeUDF);
+static NodeToPlacementGroupHashEntry * NodeToPlacementGroupHashGetNodeWithGroupId(
+	HTAB *nodePlacementGroupHash,
+	int32
+	nodeGroupId);
+
+
+/* other helpers */
+static List * PlacementListGetUniqueNodeGroupIds(List *placementList);
+static int WorkerNodeListGetNodeWithGroupId(List *workerNodeList, int32 nodeGroupId);
+
+
+/*
+ * PrepareRebalancerPlacementIsolationContext creates RebalancerPlacementIsolationContext
+ * that keeps track of which worker nodes are used to separate which shard placement groups
+ * that need separate nodes.
+ */
+RebalancerPlacementIsolationContext *
+PrepareRebalancerPlacementIsolationContext(List *activeWorkerNodeList,
+										   List *rebalancePlacementList,
+										   WorkerNode *drainWorkerNode,
+										   FmgrInfo *shardAllowedOnNodeUDF)
+{
+	HTAB *nodePlacementGroupHash =
+		CreateSimpleHashWithNameAndSize(uint32, NodeToPlacementGroupHashEntry,
+										"NodeToPlacementGroupHash",
+										list_length(activeWorkerNodeList));
+
+	activeWorkerNodeList = SortList(activeWorkerNodeList, CompareWorkerNodes);
+	rebalancePlacementList = SortList(rebalancePlacementList, CompareShardPlacements);
+
+	NodeToPlacementGroupHashInit(nodePlacementGroupHash, activeWorkerNodeList,
+								 rebalancePlacementList, drainWorkerNode);
+
+	NodeToPlacementGroupHashAssignNodes(nodePlacementGroupHash,
+										activeWorkerNodeList,
+										rebalancePlacementList,
+										shardAllowedOnNodeUDF);
+
+	RebalancerPlacementIsolationContext *context =
+		palloc(sizeof(RebalancerPlacementIsolationContext));
+	context->nodePlacementGroupHash = nodePlacementGroupHash;
```

Comment:
> Guess its current use is to hide the implemetation details from the rest of the system

Correct

> but it would be worth using the defined structure through this file, instead of passing the vanilla HTAB around.

What benefit do you see in passing the context instead? I'm ok to both but wondering why do you think we should pass the context instead of the underlying htab?

Reply:
> but it would be worth using the defined structure through this file, instead of passing the vanilla HTAB around.

done btw